### PR TITLE
JSRule: Overwrite/Update existing rule with same UID or throw error

### LINKED ISSUE #40: Feature request: Updating an existing JSRule()
Is there any way to update an existing JSRule() or to replace it with a new rule?
This would be great to update Triggers, e.g. update a cron trigger.

I would use it in one manager rule to update the triggers of a child rule or I would also be happy with just removing the child rule and then adding a new child rule.

### PR DESCRIPTION
**Pull request by florian-h05**
## Fixes #40 
@jpg0 

On creation of a new rule using `JSRule()`, the `ruleRegistry` is checked for a rule with the same UID. 

When a it already exists, throw error. 
Or:
Use the new `overwrite` parameter to overwrite an existing rule with same UID.
This can be used to dynamically update rules from other rules.

Note: Rule must be created from same file.

Signed-off-by: Florian Hotze <florianh_dev@icloud.com>

### CODE REVIEW

File: rules/rules.js
Comment (jpg0) [inline rules/rules.js:98]: I would suggest returning a boolean indicating whether a rule was actually removed or not

File: rules/rules.js
Comment (jpg0) [inline rules/rules.js:131]: I'm not sure that it's the best thing to _force_ prior removal. This means that if a script relies on failure if it tries to insert a rule a second time, it will change it's behaviour. More concretely: if two rules with the same uid are inserted, then the existing behaviour is that the second will fail. The new behaviour is that the first will be silently overwritten.

My preference would be to (a) retain the original behaviour by default (as overwriting is not a common use-case, I'd expect that doing it accidentally is more common and failing is good in this case) and (b) allow explicit overwriting by setting a parameter.

**Review by jpg0 — CHANGES_REQUESTED** [review]

File: rules/rules.js
Comment (florian-h05) [reply rules/rules.js:98]: Two things to mention:

- Removing a rule is only possible when it was created in the same file
- Returning a boolean indicatin whether it was removed or not does not work with:
```javascript
return !(ruleRegistry.remove(uid) == null);
```
It seems that null is also returned when the rule was removed, but accroding to the [Java Doc](https://www.openhab.org/javadoc/latest/org/openhab/core/common/registry/registry#remove(K)) it should return the element that was removed. It really removes the rule, but it always returns `null`.

To return whether the rule was removed, I have to use that work-around:
```javascript
if (ruleExists(uid)) {
   log.info('Removing rule: {}', ruleRegistry.get(uid).name ? ruleRegistry.get(uid).name : uid);
   ruleRegistry.remove(uid);
   return !ruleExists(uid);
} else {
   return false;
}
```

**Comment by florian-h05** [comment]
@jpg0 

I tried to bring in you requested changes. Please have a look at my comment.

File: rules/rules.js
Comment (jpg0) [reply rules/rules.js:101]: lgtm

**Review by jpg0 — APPROVED** [review]

**Comment by jpg0** [comment]
@digitaldan happy to merge this?

File: rules/rules.js
Comment (digitaldan) [inline rules/rules.js:137]: So without this option, if a rule exists with the same ID , it would thrown an error when trying to persist it?

Comment (florian-h05) [reply rules/rules.js:137]: No, it would just be quiet without that option set.

**Comment by digitaldan** [comment]
LGTM, i'm assuming I know the answer to the one question i had, so please feel free to merge !

**Review by digitaldan — APPROVED** [review]

**Comment by florian-h05** [comment]
> LGTM, i'm assuming I know the answer to the one question i had, so please feel free to merge !

When trying to add a rule with same UID and without the `override` option set, it would be quiet.

I can not merge as I have no write access.

File: rules/rules.js
Comment (digitaldan) [reply rules/rules.js:137]: huh, thats not actually what i thought would happen :-) is that desired?  i'm thinking we should always then remove the rule if it exists, or thrown an error. So right now, you would have no idea if it save properly or not without this flag right?

Comment (florian-h05) [reply rules/rules.js:137]: I don't think is desired that there is no error, but I am watching this behavious on my system.

I will add something to throw an error as overriding by default is nothing @jpg0 would do.

**Comment by florian-h05** [comment]
@digitaldan 

It now throws an error when a rule with the same UID already exists and the `override` option is not set.

File: rules/rules.js
Comment (digitaldan) [inline rules/rules.js:141]: should we instead throw the error instead of logging it?

Comment (florian-h05) [reply rules/rules.js:141]: Ups, yes, give me a few minutes.

**Comment by florian-h05** [comment]
@digitaldan 

New try :-)

**Comment by digitaldan** [comment]
LGTM, I can merge when I'm back home , unless @jpg0 you want to review and merge ?

**Comment by jpg0** [comment]
Ideally `ScriptedRuleProvider` in core should return whether a rule deletion actually occurs (instead of @florian-h05 needed to perform another check after attempting the removal), but that's too much to fix right now, so let's merge this.

**Pull request merged**
